### PR TITLE
a bit of extra logic including random zombie placement and sun points

### PR DESCRIPTION
**Pull request by bkzhang**
Read some of the comments i left in the code if you plan on editing this branch.

I added sun point functionality and split up a bit of the player logic so that invalid moves can loop more easily. Zombies generate randomly. The two variables `numZombiesToGenerate` and `numZombiesToEliminate` are different, the former when it reaches 0 stop generating zombies (not yet implemented) and the latter is that you win if it reaches 0 (since no more zombies are alive). A separate option of implementing this is to have `numZombiesToGenerate` stop generating zombies at 0, but everytime a zombie is generated, add it to a set and if `numZombiesToGenerate` is 0 and the set is empty then you win, more space but some may consider it less error prone.

Multiple zombies can exist on the same tile, peashooter always hits the first zombie on the tile. Either make it so that zombies attack plant when its on same tile (that way players can spawn a plant last minute and block the zombies) or have it so that zombies attack everything one tile to the left of them. (Zombie functionality mostly unimplemented).

### CODE REVIEW

**Comment by bkzhang** [comment]
Added the logic for moving zombies and projectiles. Currently zombie logic only works for zombies with movespeed 1. If they have a higher movespeed they may bypass certain projectiles and (maybe) plants. On the gameboard projectiles are shown with `.`, Sunflower with `S`, Peashooter with `P`, Zombie with `Z` and if both a zombie and a plant is on a single tile then its shown with `B`.

**Comment by bkzhang** [comment]
each turn's sunpoints are generated based on number of sunflowers and default turn generated sunpoints. Currently every turn you should get 5 sunpoints and sunflowers should each generate every OTHER turn 3 or so sunpoints. Defeat 10 zombies to win.

**Comment by bkzhang** [comment]
every turn the board will have a `turn()` call and each of its tiles will have a `turn()` call. The tile `turn()` method calculates how much dmg a zombie takes and how much dmg a plant takes. and the board `turn()` calculates the movement of each zombie and projectile as well as how many zombies have been eliminated and how many sunpoints are generated

File: .gitignore [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
+*.war
+*.nar
+*.ear
+*.zip
+*.tar.gz
+*.rar
+
+# virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
+hs_err_pid*
+/bin/
```
Comment (vikrambombhi) [inline .gitignore:24]: Why does it say you edited this entire file even though you only added one line?

Comment (vikrambombhi) [reply .gitignore:24]: also, why would `/bin/` ever get committed to this repo?

Comment (bkzhang) [reply .gitignore:24]: this was from @andrewdlqnguyen 's branch, was probably eclipse doing weird stuff

Comment (vikrambombhi) [reply .gitignore:24]: ok :woman_shrugging:

**Comment by vikrambombhi** [comment]
For some reason when the zombie collides with the peashooter the "merge" and just show a 'B'

**Comment by vikrambombhi** [comment]
let's merge this branch and fix the remaining problems in a separate PR. I changed one thing to get rid of a compiler warning you can read about the warning [here](http://errorprone.info/bugpattern/StaticQualifiedUsingExpression)

**Review by vikrambombhi — APPROVED** [review]

**Comment by bkzhang** [comment]
> For some reason when the zombie collides with the peashooter the "merge" and just show a 'B'

yeah that was intended, b = Both, because for now you can spawn plants ontop of the zombies, so zombies just attack whatever is in the same tile and not the one adjacent

**Pull request merged**